### PR TITLE
Added a couple of wrapper DUs for Region and VMSize.

### PR DESCRIPTION
**Pull request by isaacabraham**

### CODE REVIEW

**Comment by dsyme** [comment]
A couple of points explaining my philosophy on this:

* I should say that I have a slight aversion to revealing DUs in F# APIs (especially those that might be used from C# - see the recommendation "Do hide the representation of F# union types in vanilla .NET APIs." in the F# Component Design Guidelines).  This is because even for single-case DUs the compiled form of the API contains more logic than a class like this:

      type Region(region: string) = 
          member x.Name = region

  The compiled form of a DU contains things relevant to comparison, reflection etc.  So in the name of Occam it seems better to use a class: it reveals less and also happens to be more natural to the C# programmer.  

* However you then get questions like "should the Region type support comparison? Hashing? Equality? ToString?". This is one reason I preferred to just use a string in the API design: it avoids largely pointless questions like this in nominal type design - I mean, seriously we don't even want to be thinking about these things.  And the defaults single-case DUs are imperfect here - since they don't support ToString.

* Management code tends to be very script-like: run-once, copy-and-paste etc.  Using too many nominal types can lead to an overly typed POV. I suspect this can be a reason for Unix-like folk to reject ("ugghhh I hate all those types they get in the way"). 

* The MBrace design tends to use strings rather than new nominal types now - e.g. the removal of FileInfo in favour of strings for file names. 

So in short I'm still in favour of "keep it simple use a string"

BTW do you think we need to offer an affinity group as a region?  Or can a user already use an affinity group as a region name?

**Comment by isaacabraham** [comment]
All good points. The main concern with just strings is that it allows the caller to inadvertently mix arguments like Region and e.g. VmSize. I actually didn't think about things like equality checking etc.. I'd be happy to change this to a Class as well - it's just a bit less code for a DU and a more common way to do it in F# projects I've seen. Since we're using static properties for each region and vmsize, the user is completely oblivious to the fact that the type is now a Region instead of a String (and it will be an exceptional case when someone wants to create their own one) - indeed if you're using e.g. a region now as a string and moved over to the code in this PR, it would continue to work as before, unless you had mixed up Region and VMSize.

I'm not particularly precious about this PR to be honest - although I'm keen to hear Eirik's thoughts as well - and I agree that it needs to be lightweight for the caller - it's just a bit of extra type safety to reduce the change of them mixing things up (and there are quite a few arguments in the call site).

One thing I do think though - we should either say with strings or move to DUs/Classes - type aliases IMHO offer nothing in this situation.

**Comment by eiriktsarpalis** [comment]
I tend to agree with Don on this. 

On a more general note, I am skeptical of the utility of wrapping values using DUs (or any nominal type with accessible constructors for that matter) since it doesn't really help avoiding user errors in certain cases. For instance `VMSize "South Central US"` is an invalid expression that is well-typed. When creating such wrapper classes, I usually incorporate any validation logic in the constructor.

For the record, here's a recent related discussion on the issue: https://gist.github.com/isaacabraham/21f3d599fe9230f6bc7a

**Comment by dsyme** [comment]
Closing this - am assuming we're happy for @eiriktsarpalis to design it the way he sees fit in #124 and can discuss further there
